core/src/main/java/org/apache/iceberg/CatalogProperties.java

```diff
@@ -119,6 +119,8 @@ private CatalogProperties() {}
       "client.pool.cache.eviction-interval-ms";
   public static final long CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS_DEFAULT =
       TimeUnit.MINUTES.toMillis(5);
+  /** Name of the custom {@link ClientPool} implementation class. */
+  public static final String CLIENT_POOL_IMPL = "client-pool-impl";
```

Comment:
Is it strictly necessary to inject this using reflection? If this is for customization in a certain environment, then maybe it would be better to allow customization in a different way?

Reply:
I think the context is from https://github.com/apache/iceberg/pull/6175 (more issues scattered).  The general problem being the CachedClientPool that's always on in Iceberg Hive, which is a global cache that doesnt work in all environments, because the current key (metastoreUri) collides too much, leading to the wrong client used for a lot of use cases.

It seems simpler to me to just make cache pluggable, but I am not sure if there's a better solution I'm not seeing to make it non-dynamic?  If we dont want user to turn off CachedClientPool, at least we need to have a pluggable conf => key generator?  Or maybe just allow them to use the toggle between CachedClientPool and the raw ClientPool as an option (as a way to turn off the global cache)?  cc @pvary @flyrain  as well if any thoughts

Reply:
Perhaps another way is to allow configurable cache keys? We can have pre-defined key elements and users can use these elements to compose cache key used in `CachedClientPool`. Some key elements I can think of: HMS URI (this is probably mandatory), UGI, user name, specific configurations. When `HiveCatalog` creates `CahcedClientPool`, it can check the configured key and pass a key supplier to `CahcedClientPool`.

Reply:
Yeah I'll have a try and update the PR

Reply:
Hi @szehon-ho @rdblue , I have updated the PR to demonstrate the idea. Let me know if you think it makes sense. Thanks~

Reply:
Hi sorry, I realized I let this slip, I will try to look at this soon.